### ISSUE TITLE
500 and 503 caused by contention in D:\home\data\Functions\secrets\positions.json access.

### ISSUE THREAD

**Issue by suwatch** [Azure/azure-functions-host#2650]
This happened during scaling out of one of the function app.

```csharp
System.IO.IOException : The process cannot access the file 'D:\home\data\Functions\secrets\positions.json' because it is being used by another process.
   at System.IO.__Error.WinIOError(Int32 errorCode,String maybeFullPath)
   at System.IO.FileStream.Init(String path,FileMode mode,FileAccess access,Int32 rights,Boolean useRights,FileShare share,Int32 bufferSize,FileOptions options,SECURITY_ATTRIBUTES secAttrs,String msgPath,Boolean bFromProxy,Boolean useLongPath,Boolean checkHost)
   at System.IO.FileStream..ctor(String path,FileMode mode,FileAccess access,FileShare share,Int32 bufferSize,FileOptions options,String msgPath,Boolean bFromProxy,Boolean useLongPath,Boolean checkHost)
   at System.IO.StreamWriter.CreateFile(String path,Boolean append,Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path,Boolean append,Encoding encoding,Int32 bufferSize,Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path,Boolean append,Encoding encoding,Int32 bufferSize)
   at async Microsoft.Azure.WebJobs.Script.FileUtility.WriteAsync(String path,String contents,Encoding encoding) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script\Extensions\FileUtility.cs : 0
   at async Microsoft.Azure.WebJobs.Script.WebHost.FileSystemSecretsRepository.WriteAsync(ScriptSecretsType type,String functionName,String secretsContent) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\FileSystemSecretsRepository.cs : 0
   at async Microsoft.Azure.WebJobs.Script.WebHost.SecretManager.PersistSecretsAsync[T](T secrets,String keyScope,Boolean isNonDecryptable) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\SecretManager.cs : 0
   at async Microsoft.Azure.WebJobs.Script.WebHost.SecretManager.GetFunctionSecretsAsync(String functionName,Boolean merged) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\SecretManager.cs : 147
   at async Microsoft.Azure.WebJobs.Script.WebHost.Filters.AuthorizationLevelAttribute.GetAuthorizationResultAsync(??) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Filters\AuthorizationLevelAttribute.cs : 130
   at async Microsoft.Azure.WebJobs.Script.Host.FunctionRequestInvoker.DetermineAuthorizationLevelAsync(HttpRequestMessage request) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\FunctionRequestInvoker.cs : 121
   at async Microsoft.Azure.WebJobs.Script.Host.FunctionRequestInvoker.PreprocessRequestAsync(HttpRequestMessage request) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\FunctionRequestInvoker.cs : 108
   at async Microsoft.Azure.WebJobs.Script.WebHost.Controllers.FunctionsController.ExecuteAsync(HttpControllerContext controllerContext,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Controllers\FunctionsController.cs : 39
   at async System.Web.Http.Dispatcher.HttpControllerDispatcher.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken)
   at async System.Web.Http.Dispatcher.HttpControllerDispatcher.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken)
   at async Microsoft.Azure.WebJobs.Script.WebHost.Handlers.WebScriptHostHandler.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Handlers\WebScriptHostHandler.cs : 42
   at async Microsoft.Azure.WebJobs.Script.WebHost.Handlers.SystemTraceHandler.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Handlers\SystemTraceHandler.cs : 67
   at async System.Web.Http.HttpServer.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken)
```

```csharp
System.IO.IOException : The process cannot access the file 'D:\home\data\Functions\secrets\positions.json' because it is being used by another process.
   at System.IO.__Error.WinIOError(Int32 errorCode,String maybeFullPath)
   at System.IO.FileStream.Init(String path,FileMode mode,FileAccess access,Int32 rights,Boolean useRights,FileShare share,Int32 bufferSize,FileOptions options,SECURITY_ATTRIBUTES secAttrs,String msgPath,Boolean bFromProxy,Boolean useLongPath,Boolean checkHost)
   at System.IO.FileStream..ctor(String path,FileMode mode,FileAccess access,FileShare share,Int32 bufferSize,FileOptions options,String msgPath,Boolean bFromProxy,Boolean useLongPath,Boolean checkHost)
   at System.IO.StreamWriter.CreateFile(String path,Boolean append,Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path,Boolean append,Encoding encoding,Int32 bufferSize,Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path,Boolean append,Encoding encoding,Int32 bufferSize)
   at async Microsoft.Azure.WebJobs.Script.FileUtility.WriteAsync(String path,String contents,Encoding encoding) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script\Extensions\FileUtility.cs : 0
   at async Microsoft.Azure.WebJobs.Script.WebHost.FileSystemSecretsRepository.WriteAsync(ScriptSecretsType type,String functionName,String secretsContent) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\FileSystemSecretsRepository.cs : 124
   at async Microsoft.Azure.WebJobs.Script.WebHost.SecretManager.PersistSecretsAsync[T](T secrets,String keyScope,Boolean isNonDecryptable) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\SecretManager.cs : 422
   at async Microsoft.Azure.WebJobs.Script.WebHost.SecretManager.GetFunctionSecretsAsync(String functionName,Boolean merged) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Security\SecretManager.cs : 147
   at async Microsoft.Azure.WebJobs.Script.WebHost.Filters.AuthorizationLevelAttribute.GetAuthorizationResultAsync(??) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Filters\AuthorizationLevelAttribute.cs : 130
   at async Microsoft.Azure.WebJobs.Script.Host.FunctionRequestInvoker.DetermineAuthorizationLevelAsync(HttpRequestMessage request) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\FunctionRequestInvoker.cs : 121
   at async Microsoft.Azure.WebJobs.Script.Host.FunctionRequestInvoker.PreprocessRequestAsync(HttpRequestMessage request) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\FunctionRequestInvoker.cs : 108
   at async Microsoft.Azure.WebJobs.Script.WebHost.Controllers.FunctionsController.ExecuteAsync(HttpControllerContext controllerContext,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Controllers\FunctionsController.cs : 39
   at async System.Web.Http.Dispatcher.HttpControllerDispatcher.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken)
   at async System.Web.Http.Dispatcher.HttpControllerDispatcher.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken)
   at async Microsoft.Azure.WebJobs.Script.WebHost.Handlers.WebScriptHostHandler.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Handlers\WebScriptHostHandler.cs : 42
   at async Microsoft.Azure.WebJobs.Script.WebHost.Handlers.SystemTraceHandler.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-script\src\WebJobs.Script.WebHost\Handlers\SystemTraceHandler.cs : 67
   at async System.Web.Http.HttpServer.SendAsync(HttpRequestMessage request,CancellationToken cancellationToken)
```

**Comment by davidebbo** [comment]
So presumably it was trying to write the secret file because this was the first access to the function and it had not been created yet? I would think that after it exists, it should be all reads and no writes (unless user explicitly resets the secrets).

**Comment by suwatch** [comment]
I could not find any reset secret api call (assuming SCM /api/functions) during 6 hours (between those runs). Looking at the site, currently `D:\home\data\Functions\secrets\positions.json` file does not exist.    Nor is there a function called `positions` under `d:\home\site\wwwroot` physically.   It is a proxy function making outgoing call to another function app.   Hopefully this yields more clue.

**Comment by fabiocav** [comment]
Closed by #2695
